src/circuit/gadget/ecc/chip/add.rs

```diff
@@ -0,0 +1,469 @@
+
+            // (1 - x_q * γ) * (y_r - y_p) = 0
+            let poly10 = (one.clone() - if_gamma) * (y_r.clone() - y_p);
+
+            // ((1 - (x_q - x_p) * α - (y_q + y_p) * δ)) * x_r
+            let poly11 = (one.clone() - if_alpha.clone() - if_delta.clone()) * x_r;
+
+            // ((1 - (x_q - x_p) * α - (y_q + y_p) * δ)) * y_r
+            let poly12 = (one - if_alpha - if_delta) * y_r;
+
+            [
+                poly1, poly2, poly3, poly4, poly5, poly6, poly7, poly8, poly9, poly10, poly11,
+                poly12,
+            ]
+            .iter()
+            .map(|poly| q_add.clone() * poly.clone())
+            .collect()
+        });
+    }
+
+    pub(super) fn assign_region<C: CurveAffine>(
+        &self,
+        p: &EccPoint<C>,
+        q: &EccPoint<C>,
+        offset: usize,
+        region: &mut Region<'_, C::Base>,
+    ) -> Result<EccPoint<C>, Error> {
+        // Enable `q_add` selector
+        self.q_add.enable(region, offset)?;
+
+        // Copy point `p` into `x_p`, `y_p` columns
+        copy(region, || "x_p", self.x_p, offset, &p.x, &self.perm)?;
+        copy(region, || "y_p", self.y_p, offset, &p.y, &self.perm)?;
+
+        // Copy point `q` into `x_qr`, `y_qr` columns
+        copy(region, || "x_q", self.x_qr, offset, &q.x, &self.perm)?;
+        copy(region, || "y_q", self.y_qr, offset, &q.y, &self.perm)?;
+
+        let (x_p, y_p) = (p.x.value(), p.y.value());
+        let (x_q, y_q) = (q.x.value(), q.y.value());
+
+        // Assign α = inv0(x_q - x_p)
+        let alpha = x_p.zip(x_q).map(|(x_p, x_q)| inv0(x_q - x_p));
+        region.assign_advice(
+            || "α",
+            self.alpha,
+            offset,
+            || alpha.ok_or(Error::SynthesisError),
+        )?;
+
+        // Assign β = inv0(x_p)
+        region.assign_advice(
+            || "β",
+            self.beta,
+            offset,
+            || {
+                let beta = x_p.map(inv0);
+                beta.ok_or(Error::SynthesisError)
+            },
+        )?;
+
+        // Assign γ = inv0(x_q)
+        region.assign_advice(
+            || "γ",
+            self.gamma,
+            offset,
+            || {
+                let gamma = x_q.map(inv0);
+                gamma.ok_or(Error::SynthesisError)
+            },
+        )?;
+
+        // Assign δ = inv0(y_q + y_p) if x_q = x_p, 0 otherwise
+        region.assign_advice(
+            || "δ",
+            self.delta,
+            offset,
+            || {
+                let x_p = x_p.ok_or(Error::SynthesisError)?;
+                let x_q = x_q.ok_or(Error::SynthesisError)?;
+                let y_p = y_p.ok_or(Error::SynthesisError)?;
+                let y_q = y_q.ok_or(Error::SynthesisError)?;
+
+                let delta = if x_q == x_p {
+                    inv0(y_q + y_p)
+                } else {
+                    C::Base::zero()
+                };
+                Ok(delta)
+            },
+        )?;
```

Comment:
Unresolved comment: https://github.com/zcash/orchard/pull/73#discussion_r645096894

Reply:
Batch inversion for `alpha`, `beta`, `gamma`, `delta` implemented in https://github.com/zcash/orchard/pull/107/commits/4a0f7bad4e3fc94ddc94d0bf92f6b52b257d2840.